packages/lexicon/src/validators/primitives.ts

```diff
@@ -248,39 +246,76 @@ export function string(
     }
   }
 
+  if (typeof def.format === 'string') {
+    switch (def.format) {
+      case 'datetime':
+        return datetimeFormat(path, value)
+      case 'at-uri':
+        return atUriFormat(path, value)
+      case 'did':
+        return didFormat(path, value)
+      case 'cid':
+        return cidFormat(path, value)
+    }
+  }
+
   return { success: true, value }
 }
 
-export function datetime(
-  lexicons: Lexicons,
-  path: string,
-  def: LexUserType,
-  value: unknown,
-): ValidationResult {
-  def = def as LexDatetime
+export function datetimeFormat(path: string, value: string): ValidationResult {
+  try {
+    if (!isValidISODateString(value)) {
+      throw new Error()
+    }
+  } catch {
+    return {
+      success: false,
+      error: new ValidationError(
+        `${path} must be an iso8601 formatted datetime`,
+      ),
+    }
+  }
+  return { success: true, value }
+}
 
-  // type
-  const type = typeof value
-  if (type !== 'string') {
+export function atUriFormat(path: string, value: string): ValidationResult {
+  try {
+    const uri = new AtUri(value)
+    if (!value.startsWith('at://') || uri.toString() !== value) {
+      throw new Error()
+    }
```

Comment:
I am not sure if we need the `uri.toString() !== value` check.  Two reasons there:
 - The `AtUri` constructor should throw on an invalid uri.
 - It might create false positives since there isn't necessarily a canonical representation of a URI (e.g. trailing slashes, empty search params or hash fragments, unnecessary uri encoding).

Just to toss a possibility out there:
```suggestion
    if (!value.startsWith('at://')) {
      throw new Error() // Fast path
    }
    const uri = new AtUri(value)
    if (!uri.host) {
      throw new Error()
    }
```

Reply:
yup good point 👍 

i'll be merging in bryans identifier validators soon: https://github.com/bluesky-social/atproto/pull/576